packages/flutter/lib/src/widgets/actions.dart

```diff
@@ -276,88 +520,169 @@ class Actions extends InheritedWidget {
-            '$Actions.invoke() was called with a context that does not contain an '
-            '$Actions widget.\n'
-            'No $Actions ancestor could be found starting from the context that '
-            'was passed to $Actions.invoke(). This can happen if the context comes '
-            'from a widget above those widgets.\n'
-            'The context used was:\n'
-            '  $context');
-      }
       if (action == null) {
-        throw FlutterError('Unable to find an action for an intent in the $Actions widget in the context.\n'
-            "$Actions.invoke() was called on an $Actions widget that doesn't "
-            'contain a mapping for the given intent.\n'
+        throw FlutterError('Unable to find an action for an Intent with type '
+            '${intent.runtimeType} in an $Actions widget in the given context.\n'
+            '$Actions.invoke() was unable to find an $Actions widget that '
+            "contained a mapping for the given intent, or the intent type isn't the "
+            'same as the type argument to invoke (which is $T - try supplying a '
+            'type argument to invoke if one was not given)\n'
             'The context used was:\n'
             '  $context\n'
-            'The intent requested was:\n'
-            '  $intent');
+            'The intent type requested was:\n'
+            '  ${intent.runtimeType}');
       }
       return true;
     }());
-    if (action == null) {
-      // Will only get here if nullOk is true.
-      return false;
-    }
-
-    // Invoke the action we found using the dispatcher from the Actions Element
-    // we found, using the given focus node.
-    return _findDispatcher(actionsElement).invokeAction(action, intent, focusNode: focusNode);
+    // Invoke the action we found using the relevant dispatcher from the Actions
+    // Element we found.
+    return actionElement != null ? _findDispatcher(actionElement).invokeAction(action, intent, context) != null : null;
   }
 
   @override
-  bool updateShouldNotify(Actions oldWidget) {
-    return oldWidget.dispatcher != dispatcher || !mapEquals<LocalKey, ActionFactory>(oldWidget.actions, actions);
-  }
+  State<Actions> createState() => _ActionsState();
 
   @override
   void debugFillProperties(DiagnosticPropertiesBuilder properties) {
     super.debugFillProperties(properties);
     properties.add(DiagnosticsProperty<ActionDispatcher>('dispatcher', dispatcher));
-    properties.add(DiagnosticsProperty<Map<LocalKey, ActionFactory>>('actions', actions));
+    properties.add(DiagnosticsProperty<Map<Type, Action<Intent>>>('actions', actions));
+  }
+}
+
+class _ActionsState extends State<Actions> {
+  // Keeps the last-known enabled state of each action in the action map.
+  Map<Action<Intent>, bool> enabledState = <Action<Intent>, bool>{};
+  // Used to tell the market to rebuild when the enabled state of an action in
+  // the map changes.
+  LocalKey rebuildKey = UniqueKey();
```

Comment:
Why is this a Key? Seems like its not really used as a key for a widget? Maybe just use an Object as marker?